src/quadratic_newton.jl

```diff
@@ -0,0 +1,217 @@
+
+"""
+    Semi-smooth Newton method (Algorithm 2 of Lorenz et al. 2019) for solving quadratically regularised optimal transport
+
+    See also: [`QuadraticOTNewton`](@ref), [`quadreg`](@ref)
+
+    # References
+    
+    Lorenz, Dirk A., Paul Manns, and Christian Meyer. [Quadratically regularized optimal transport.](https://doi.org/10.1007/s00245-019-09614-w) Applied Mathematics & Optimization 83.3 (2021): 1919-1949.
+"""
+struct QuadraticOTNewton{T<:Real,K<:Real,D<:Real} <: QuadraticOT
+    θ::T
+    κ::K
+    δ::D
+end
+
+"""
+    QuadraticOTNewton(θ = 0.1, κ = 0.5, δ = 1e-5)
+
+    Semi-smooth Newton method (Algorithm 2 of Lorenz et al. 2019) with Armijo parameters `θ` and `κ`, and conjugate gradient regularisation `δ`. 
+    
+    See also: [`QuadraticOTNewton`](@ref), [`quadreg`](@ref)
+"""
+function QuadraticOTNewton(θ=0.1, κ=0.5, δ=1e-5)
+    return QuadraticOTNewton(θ, κ, δ)
+end
+
+Base.show(io::IO, ::QuadraticOTNewton) = print(io, "Semi-smooth Newton algorithm")
+
+struct QuadraticOTNewtonCache{U,V,C,P,GT,X}
+    u::U
+    v::V
+    δu::U
+    δv::V
+    σ::C
+    γ::P
+    G::GT
+    x::X
+    M::Int
+    N::Int
+end
+
+function build_cache(
+    ::Type{T},
+    ::QuadraticOTNewton,
+    μ::AbstractVector,
+    ν::AbstractVector,
+    C::AbstractMatrix,
+    ε::Real,
+) where {T}
+    # create and initialize dual potentials
+    u = similar(μ, T, size(μ, 1))
+    v = similar(ν, T, size(ν, 1))
+    fill!(u, zero(T))
+    fill!(v, zero(T))
+    δu = zeros(T, size(u))
+    δv = zeros(T, size(v))
+    # intermediate variables (don't need to be initialised)
+    σ = similar(C, T)
+    γ = similar(C, T)
+    M = size(μ, 1)
+    N = size(ν, 1)
+    G = zeros(T, M + N, M + N)
+    # initial guess for conjugate gradient 
+    x = zeros(T, M + N)
+    return QuadraticOTNewtonCache(u, v, δu, δv, σ, γ, G, x, M, N)
+end
+
+function check_convergence(
+    μ::AbstractVector,
+    ν::AbstractVector,
+    cache::QuadraticOTNewtonCache,
+    convergence_cache::QuadraticOTConvergenceCache,
+    atol::Real,
+    rtol::Real,
+)
+    γ = cache.γ
+    norm_diff = max(
+        norm(vec(sum(γ; dims=2)) .- μ, Inf), norm(vec(sum(γ; dims=1) .- ν), Inf)
+    )
+    isconverged =
+        norm_diff <
+        max(atol, rtol * max(convergence_cache.norm_source, convergence_cache.norm_target))
+    return isconverged, norm_diff
+end
+
+function descent_dir!(solver::QuadraticOTSolver{<:QuadraticOTNewton})
+    # unpack solver
+    eps = solver.eps
+    C = solver.C
+    μ = solver.source
+    ν = solver.target
+    cache = solver.cache
+    # unpack cache
+    u = cache.u
+    v = cache.v
+    δu = cache.δu
+    δv = cache.δv
+    σ = cache.σ
+    γ = cache.γ
+    G = cache.G
+    x = cache.x
+    M = cache.M
+    N = cache.N
+    # Armijo parameters
+    δ = solver.alg.δ
+
+    # setup intermediate variables 
+    @. γ = u + v' - C
+    @. σ = ifelse(γ ≥ 0, 1, 0)
```

Comment:
Do we need integers here? I.e., could we use
```suggestion
    @. σ = γ ≥ 0
```
?